File: Server/steerlab_server/experiment/evaluate_subsample.py
"""Seeded, stratified subsampling for the per-response coding evaluate —
the line-for-line twin of ``Sources/ExperimentKit/EvaluateSubsample.swift``.

**Why this module exists (field discovery 2026-08-29).** A 7,200-record
corpus needed judged classification and the preregistered design was a
stratified 2,400-record subsample. ``evaluate`` codes a whole source run and
nothing else, so the operation had no honest spelling: the only routes
available were coding all 7,200 (not the declared design, and far more judge
calls than the power computation asked for) or hand-building a run directory
holding the chosen records — which is evidence-chain corruption under
immutable ``runs/`` and was correctly refused. The design was legitimate and
preregistered; the instrument simply could not say it. This is the sixth gap
of that shape.

**The contract.** ``--sample-per-condition <n>`` and ``--sample-seed
<hex-or-int>`` are BOTH given or NEITHER. A sample without a seed is a
subsample nobody can redraw; a seed without a sample size is a stamp on a
coding it did not shape. Either half alone refuses at 64 — no defaulted seed,
no inferred size. There is deliberately no fraction spelling: an absolute
per-condition ``n`` is what a power computation produces, and "40% of
whatever happened to be there" is not a design.

**Never clamp.** ``n`` larger than a condition's codeable population refuses
(64). Clamping would code a smaller design than the preregistered one while
every stamp said ``samplePerCondition: n`` — the silent substitution the
whole refusal vocabulary exists to prevent.

**The draw is the house RNG.** Every choice here is a partial Fisher–Yates
over :class:`~steerlab_server.steering.token_bank_downsampling.SplitMix64`,
seeded through SHA-256, for exactly the reason the neutral token bank adopted
it (2026-08-28 audit, convention note 9): ``random.sample`` is allowed to
change algorithm across CPython versions, and a subsample whose membership
can move under a runtime upgrade while its seed stamp stays identical is not
reproducible evidence. Nothing here touches the standard library's RNG, and
:mod:`tests.test_evaluate_subsample` pins the output with integer literals
that appear byte-identically in ``EvaluateSubsampleTests.swift``.

**Ordering is by UTF-8 bytes, on purpose.** Swift's ``<`` on ``String`` is
Unicode canonical-equivalence-aware and Python's ``sorted`` is code-point
order; the two agree on ASCII and are not guaranteed to agree beyond it. The
draw's stratum order is load-bearing for WHICH records are chosen, so both
engines sort promptIDs (and conditions) by their UTF-8 byte sequences.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass

from ..steering.token_bank_downsampling import SplitMix64

_U64 = (1 << 64) - 1

#: The derivation, stated once and stamped VERBATIM into every sampled
#: coding report and run config. A reader who has the seed, the source run and
#: this string can recompute the membership of the subsample by hand; a reader
#: who has a report from a future version can tell whether the rule moved,
#: because the version marker moves with it.
RULE = (
    "stratifiedByPromptID/v1 — within each condition, floor(n / P) records "
    "per promptID over that condition's P promptIDs, the n mod P remainder "
    "given one at a time to promptIDs in seeded order (a promptID already at "
    "its codeable population is skipped and its quota passes on, so exactly "
    "n records are always drawn); within each (condition, promptID) cell the "
    "records are drawn over sampleIndex ascending. Every draw is a partial "
    "Fisher-Yates over SplitMix64 seeded from the first 8 bytes, big-endian, "
    "of SHA-256(seed as 8-byte big-endian || each part length-prefixed as an "
    "8-byte big-endian UTF-8 byte count followed by those bytes): the parts "
    "are (condition,) for the promptID order and (condition, promptID) for a "
    "cell draw. promptIDs are ordered by UTF-8 bytes. Kept records stay in "
    "their source-run order."
)


class SubsampleRefusal(ValueError):
    """A typed refusal carrying its machine code and a runnable repair.

    Every member of this family is exit 64 (``blocked``): the invocation
    describes a design the verb cannot execute, which is a malformed ask
    rather than a gate declining a healthy study. The CLI maps ``code``
    through its refusal table exactly as ``vectors mirror-poles`` maps
    ``PoleMirrorError.kind``.
    """

    def __init__(self, reason: str, *, code: str, repair: str) -> None:
        super().__init__(reason)
        self.reason = reason
        self.code = code
        self.repair_action = repair


@dataclass(frozen=True)
class SubsampleRequest:
    """A validated ``(n, seed)`` ask. ``seed_text`` is the canonical
    ``0x``-prefixed 16-hex-digit spelling that every stamp carries — JSON has
    no unsigned 64-bit integer, and a decimal that a reader's JSON parser
    rounds is a seed that no longer redraws its own subsample."""

    sample_per_condition: int
    seed: int

    @property
    def seed_text(self) -> str:
        return format_seed(self.seed)


def format_seed(seed: int) -> str:
    """The canonical seed spelling: ``0x`` + 16 lowercase hex digits."""
    return f"0x{int(seed) & _U64:016x}"


def parse_seed(text: str, *, program: str) -> int:
    """``--sample-seed`` as a 64-bit unsigned integer.

    Accepts ``0x``-prefixed hex, a bare decimal, or bare hex (which a decimal
    string is read as decimal first — ``1234`` is 1234, not 0x1234, because a
    seed a researcher typed as a number must mean the number they typed).
    Anything else, and anything that does not fit in 64 bits, refuses: a
    silently truncated seed stamps a value that does not redraw its own
    subsample.
    """
    raw = (text or "").strip()
    if not raw:
        raise SubsampleRefusal(
            "--sample-seed is empty — a subsample's seed is the only thing "
            "that lets anyone redraw it, so it cannot be blank",
            code="sampleSeedMalformed",
            repair=_seed_repair(program))
    body = raw
    base = 10
    if len(body) > 2 and body[:2].lower() == "0x":
        body, base = body[2:], 16
    elif not body.isdigit():
        base = 16
    # ONE sentence for every way a seed can be unusable — not a number, a
    # negative, or too wide to represent. Swift's `UInt64(_:radix:)` cannot
    # tell those apart, and a refusal a researcher reads on one engine but
    # not the other is worse than a coarser taxonomy: twin sentences beat a
    # finer classification only one engine can make.
    value: int | None = None
    if body:
        try:
            value = int(body, base)
        except ValueError:
            value = None
    if value is None or value < 0 or value > _U64:
        raise SubsampleRefusal(
            f"--sample-seed '{raw}' is not a 64-bit unsigned number — a seed "
            "is a decimal integer, or hexadecimal with or without a '0x' "
            "prefix, of at most 16 hex digits (the leading 16 of a digest "
            "are a fine seed, written down as such)",
            code="sampleSeedMalformed",
            repair=_seed_repair(program))
    return value


def _seed_repair(program: str) -> str:
    return (f"{program} experiment evaluate <name> --sample-per-condition "
            "<n> --sample-seed 0x5eed0a5e5eed0a5e  (any 64-bit value; record "
            "it in the preregistration — the same seed always draws the same "
            "records)")


def resolve_request(sample_per_condition, sample_seed, *,
                    program: str) -> SubsampleRequest | None:
    """Validate the flag PAIR before anything is read or written.

    Returns ``None`` for the full-corpus case (neither flag given), a
    validated request when both are, and refuses when exactly one is. This
    runs at the CLI edge AND at the top of the task, so a library caller
    cannot reach the draw with half a request.
    """
    sizes = ("" if sample_per_condition is None
             else str(sample_per_condition).strip())
    seeds = "" if sample_seed is None else str(sample_seed).strip()
    if not sizes and not seeds:
        return None
    if sizes and not seeds:
        raise SubsampleRefusal(
            f"--sample-per-condition {sizes} was given without "
            "--sample-seed: a subsample nobody can redraw is not evidence, "
            "so the draw refuses rather than choosing a seed for you",
            code="sampleSeedMissing",
            repair=_seed_repair(program))
    if seeds and not sizes:
        return _refuse_seed_without_size(seeds, program)
    try:
        count = int(sizes)
    except ValueError:
        count = 0
    if count < 1:
        raise SubsampleRefusal(
            "--sample-per-condition must be a whole number of records of at "
            f"least 1, not '{sizes}' — a subsample of zero records is a "
            "design nobody can report",
            code="sampleSizeMalformed",
            repair=(f"{program} experiment evaluate <name> "
                    "--sample-per-condition 2400 --sample-seed <seed>, or "
                    "drop both flags to code the full corpus"))
    return SubsampleRequest(count, parse_seed(seeds, program=program))


def _refuse_seed_without_size(seeds: str, program: str):
    raise SubsampleRefusal(
        f"--sample-seed {seeds} was given without "
        "--sample-per-condition: with no sample size the full corpus is "
        "coded, and the seed would be stamped on a coding it did not shape",
        code="sampleSizeMissing",
        repair=(f"add --sample-per-condition <n> to draw a subsample, or "
                f"drop --sample-seed and run {program} experiment evaluate "
                "<name> to code the full corpus"))


def stream_seed(seed: int, *parts: str) -> int:
    """The per-stratum stream seed: first 8 bytes, big-endian, of
    SHA-256(seed as 8-byte big-endian ‖ each length-prefixed UTF-8 part).

    Length-prefixing is not decoration: without it ``("ab", "c")`` and
    ``("a", "bc")`` would hash identically, so two different cells of the
    same condition could share a draw. Same reason
    ``token_bank_downsampling.corpus_hash`` length-prefixes its texts.
    """
    digest = hashlib.sha256()
    digest.update((int(seed) & _U64).to_bytes(8, "big"))
    for part in parts:
        encoded = part.encode("utf-8")
        digest.update(len(encoded).to_bytes(8, "big"))
        digest.update(encoded)
    return int.from_bytes(digest.digest()[:8], "big")


def seeded_order(count: int, seed: int) -> list[int]:
    """A full Fisher–Yates permutation of ``range(count)`` over SplitMix64 —
    the same loop as ``token_bank_downsampling.selected_indices``, without the
    final sort, because here the ORDER is the answer (which promptID receives
    the next remainder record)."""
    count = int(count)
    if count <= 0:
        return []
    rng = SplitMix64(seed)
    pool = list(range(count))
    for position in range(count):
        remaining = count - position
        other = position + rng.next() % remaining
        pool[position], pool[other] = pool[other], pool[position]
    return pool


def selected_positions(count: int, keep: int, seed: int) -> list[int]:
    """The ``keep`` positions to KEEP out of ``count``, sorted ascending —
    ``token_bank_downsampling.selected_indices`` under the name this module
    reads it by. Delegated rather than re-implemented so a single loop
    defines every draw in the system."""
    from ..steering.token_bank_downsampling import selected_indices
    return selected_indices(count, keep, seed)


def _utf8_key(text: str) -> bytes:
    return (text or "").encode("utf-8")


def _condition_of(record: dict) -> str:
    return str(record.get("condition") or "")


def _prompt_of(record: dict) -> str:
    return str(record.get("promptID") or "")


def _sample_index_of(record: dict) -> int:
    try:
        return int(record.get("sampleIndex") or 0)
    except (TypeError, ValueError):
        return 0


def population_by_condition(records: list[dict]) -> dict[str, int]:
    """Codeable records per condition — what an over-ask is measured against."""
    counts: dict[str, int] = {}
    for record in records:
        condition = _condition_of(record)
        counts[condition] = counts.get(condition, 0) + 1
    return counts


def select(records: list[dict], request: SubsampleRequest, *,
           program: str) -> tuple[list[dict], dict]:
    """Draw the subsample. Returns ``(kept records in source order, stamp)``.

    Refuses — writing nothing, because this runs before the evaluate run
    directory is minted — when any condition holds fewer than
    ``samplePerCondition`` codeable records.
    """
    population = population_by_condition(records)
    short = sorted(
        ((condition, size) for condition, size in population.items()
         if size < request.sample_per_condition),
        key=lambda item: (item[1], _utf8_key(item[0])))
    if short:
        tightest, available = short[0]
        named = ", ".join(f"'{c}' has {n}" for c, n in short)
        raise SubsampleRefusal(
            f"--sample-per-condition {request.sample_per_condition} exceeds "
            f"what the source run holds: {named} codeable record(s). A "
            "subsample cannot be larger than the stratum it is drawn from, "
            "and clamping it would code a smaller design than the one that "
            "was preregistered while every stamp still said "
            f"{request.sample_per_condition}",
            code="samplePopulation",
            repair=(f"re-run with --sample-per-condition {available} or less "
                    f"(condition '{tightest}' is the binding stratum), or "
                    f"drop both sample flags and run {program} experiment "
                    f"evaluate <name> to code all {len(records)} record(s)"))

    keep: set[int] = set()
    for condition in sorted(population, key=_utf8_key):
        cells: dict[str, list[int]] = {}
        for ordinal, record in enumerate(records):
            if _condition_of(record) != condition:
                continue
            cells.setdefault(_prompt_of(record), []).append(ordinal)
        prompts = sorted(cells, key=_utf8_key)
        for ordinals in cells.values():
            ordinals.sort(key=lambda o: (_sample_index_of(records[o]), o))
        allot = _allotments(
            [len(cells[p]) for p in prompts], request.sample_per_condition,
            stream_seed(request.seed, condition))
        for prompt, quota in zip(prompts, allot):
            ordinals = cells[prompt]
            for position in selected_positions(
                    len(ordinals), quota,
                    stream_seed(request.seed, condition, prompt)):
                keep.add(ordinals[position])

    kept = [record for ordinal, record in enumerate(records) if ordinal in keep]
    return kept, stamp(request, sampled=len(kept), source=len(records))


def _allotments(populations: list[int], total: int, seed: int) -> list[int]:
    """How many records each promptID contributes: ``floor(total / P)`` each,
    then the remainder handed out one at a time in seeded promptID order,
    skipping any promptID already at its population and passing its quota on.

    The skip is what makes "exactly ``total`` records, always" an invariant on
    a ragged run (a partial or resumed source run need not be rectangular).
    It terminates because :func:`select` has already refused when the
    condition's total population is below ``total``, so headroom always
    exists until the last record is placed.
    """
    count = len(populations)
    if count == 0 or total <= 0:
        return [0] * count
    allot = [min(total // count, size) for size in populations]
    order = seeded_order(count, seed)
    remaining = total - sum(allot)
    while remaining > 0:
        placed = 0
        for index in order:
            if remaining == 0:
                break
            if allot[index] < populations[index]:
                allot[index] += 1
                remaining -= 1
                placed += 1
        if placed == 0:  # pragma: no cover - guarded by the population refusal
            break
    return allot


def stamp(request: SubsampleRequest, *, sampled: int, source: int) -> dict:
    """The ``sampling`` block. Additive by construction: its ABSENCE is what
    a full-corpus coding looks like, so every report written before this
    existed reads back byte-identically and no reader has to interpret a
    missing block as anything but "all of it"."""
    return {
        "rule": RULE,
        "samplePerCondition": request.sample_per_condition,
        "sampleSeed": request.seed_text,
        "sampledRecords": int(sampled),
        "sourceRecords": int(source),
    }


def coded_phrase(stamp_block: dict | None, total: int) -> str:
    """The human count every line says: ``"7200 record(s)"`` for a full
    corpus, ``"2400 of 7200 record(s) (seeded subsample)"`` for a sampled one.

    One function so no line can drift into implying a full-corpus coding —
    the whole point of the loud stamping is that a reader CANNOT mistake the
    two, and a report is read line by line, not block by block.
    """
    if not stamp_block:
        return f"{total} record(s)"
    return (f"{stamp_block['sampledRecords']} of "
            f"{stamp_block['sourceRecords']} record(s) (seeded subsample)")


def paired_refusal(program: str) -> SubsampleRefusal:
    """The sample flags on a pairedJudge evaluate.

    Scoped deliberately (2026-08-29): the coding instrument's unit of analysis
    is a RECORD, which is what a per-condition stratified draw is defined
    over. The paired judge's unit is a PAIR — a variant record joined to the
    baseline record of its (promptID, sampleIndex) cell — where "baseline" is
    not a sampled condition at all but the other half of every comparison, so
    "n records per condition" does not name a set of pairs. Silently ignoring
    the flags would be the trap the ``--shard`` refusal exists to prevent: a
    correct-looking command line the CLI half-executes.
    """
    return SubsampleRefusal(
        "--sample-per-condition/--sample-seed apply to the per-response "
        "coding instrument only: this study's pinned rubric is a paired "
        "comparison, whose unit is a (baseline, variant) PAIR rather than a "
        "record, so a per-condition record count does not name a set of "
        "pairs to judge",
        code="sampleUnsupportedInstrument",
        repair=(f"drop both sample flags and run {program} experiment "
                "evaluate <name> to judge every pair, or pin a "
                "perResponseCoding rubric if per-record coding is the design"))

File: Sources/ExperimentKit/EvaluateSubsample.swift
import CryptoKit
import Foundation
import SteeringKit

/// Seeded, stratified subsampling for the per-response coding evaluate —
/// the line-for-line twin of
/// `Server/steerlab_server/experiment/evaluate_subsample.py`.
///
/// **Why this exists (field discovery 2026-08-29).** A 7,200-record corpus
/// needed judged classification and the preregistered design was a stratified
/// 2,400-record subsample. `evaluate` codes a whole source run and nothing
/// else, so the operation had no honest spelling: the only routes available
/// were coding all 7,200 (not the declared design, and far more judge calls
/// than the power computation asked for) or hand-building a run directory
/// holding the chosen records — which is evidence-chain corruption under
/// immutable `runs/` and was correctly refused. The design was legitimate and
/// preregistered; the instrument simply could not say it.
///
/// **The contract.** `--sample-per-condition <n>` and `--sample-seed
/// <hex-or-int>` are BOTH given or NEITHER. A sample without a seed is a
/// subsample nobody can redraw; a seed without a sample size is a stamp on a
/// coding it did not shape. Either half alone is a malformed invocation (exit
/// 64) — no defaulted seed, no inferred size. There is deliberately no
/// fraction spelling: an absolute per-condition `n` is what a power
/// computation produces.
///
/// **Never clamp.** An `n` larger than a condition's codeable population
/// refuses. Clamping would code a smaller design than the preregistered one
/// while every stamp still said `samplePerCondition: n`.
///
/// **The draw is the house RNG.** Every choice is a partial Fisher–Yates over
/// `SplitMix64`, seeded through SHA-256 — the same primitive and the same
/// argument as `TokenBankDownsampler` (2026-08-28 audit, convention note 9):
/// a subsample whose membership can move under a runtime upgrade while its
/// seed stamp stays identical is not reproducible evidence.
/// `EvaluateSubsampleTests` pins the output with integer literals that appear
/// byte-identically in `Server/tests/test_evaluate_subsample.py`.
///
/// **Ordering is by UTF-8 bytes, on purpose.** Swift's `<` on `String` is
/// Unicode canonical-equivalence-aware and Python's `sorted` is code-point
/// order; the two agree on ASCII and are not guaranteed to agree beyond it.
/// The stratum order decides WHICH records are chosen, so both engines sort
/// promptIDs (and conditions) by their UTF-8 byte sequences.
public enum EvaluateSubsample {

    /// The derivation, stated once and stamped VERBATIM into every sampled
    /// coding report and run config. A reader who has the seed, the source
    /// run and this string can recompute the subsample's membership by hand;
    /// the version marker moves if the rule ever does. Twin literal:
    /// `evaluate_subsample.RULE`.
    public static let rule = """
        stratifiedByPromptID/v1 — within each condition, floor(n / P) records \
        per promptID over that condition's P promptIDs, the n mod P remainder \
        given one at a time to promptIDs in seeded order (a promptID already \
        at its codeable population is skipped and its quota passes on, so \
        exactly n records are always drawn); within each (condition, \
        promptID) cell the records are drawn over sampleIndex ascending. \
        Every draw is a partial Fisher-Yates over SplitMix64 seeded from the \
        first 8 bytes, big-endian, of SHA-256(seed as 8-byte big-endian || \
        each part length-prefixed as an 8-byte big-endian UTF-8 byte count \
        followed by those bytes): the parts are (condition,) for the promptID \
        order and (condition, promptID) for a cell draw. promptIDs are \
        ordered by UTF-8 bytes. Kept records stay in their source-run order.
        """

    // MARK: - The request

    /// A validated `(n, seed)` ask. `seedText` is the canonical
    /// `0x`-prefixed 16-hex-digit spelling that every stamp carries — JSON
    /// has no unsigned 64-bit integer, and a decimal a reader's JSON parser
    /// rounds is a seed that no longer redraws its own subsample.
    public struct Request: Sendable, Equatable {
        public let samplePerCondition: Int
        public let seed: UInt64

        public init(samplePerCondition: Int, seed: UInt64) {
            self.samplePerCondition = samplePerCondition
            self.seed = seed
        }

        public var seedText: String { EvaluateSubsample.format(seed: seed) }
    }

    /// The `sampling` block. Additive by construction: its ABSENCE is what a
    /// full-corpus coding looks like, so every report written before this
    /// existed reads back byte-identically and no reader has to interpret a
    /// missing block as anything but "all of it".
    public struct Stamp: Codable, Sendable, Equatable {
        public let rule: String
        public let samplePerCondition: Int
        public let sampleSeed: String
        public let sampledRecords: Int
        public let sourceRecords: Int

        /// The same five keys as a `[String: Any]`, for `RunMetadata`'s
        /// JSONSerialization payload — which takes a plain object rather
        /// than an `Encodable`.
        public var jsonObject: [String: Any] {
            [
                "rule": rule,
                "samplePerCondition": samplePerCondition,
                "sampleSeed": sampleSeed,
                "sampledRecords": sampledRecords,
                "sourceRecords": sourceRecords,
            ]
        }
    }

    /// The canonical seed spelling: `0x` + 16 lowercase hex digits.
    public static func format(seed: UInt64) -> String {
        String(format: "0x%016lx", seed)
    }

    /// `--sample-seed` as a 64-bit unsigned integer.
    ///
    /// Accepts `0x`-prefixed hex, a bare decimal, or bare hex — an
    /// all-digits string is read as DECIMAL, because a seed a researcher
    /// typed as a number must mean the number they typed. Anything else, and
    /// anything that does not fit in 64 bits, refuses: a silently truncated
    /// seed stamps a value that does not redraw its own subsample.
    public static func parseSeed(_ text: String, program: String) throws -> UInt64 {
        let raw = text.trimmingCharacters(in: .whitespaces)
        guard !raw.isEmpty else {
            throw ExperimentError.malformed(
                "--sample-seed is empty — a subsample's seed is the only "
                    + "thing that lets anyone redraw it, so it cannot be blank",
                repair: seedRepair(program: program))
        }
        var body = raw
        var radix = 10
        if body.count > 2, body.prefix(2).lowercased() == "0x" {
            body = String(body.dropFirst(2))
            radix = 16
        } else if !body.allSatisfy({ $0.isNumber }) {
            radix = 16
        }
        guard !body.isEmpty, let value = UInt64(body, radix: radix) else {
            // `UInt64(_:radix:)` fails identically for a non-number and for a
            // value too large to represent, so the message names both: the
            // repair is the same sentence either way.
            throw ExperimentError.malformed(
                "--sample-seed '\(raw)' is not a 64-bit unsigned number — a "
                    + "seed is a decimal integer, or hexadecimal with or "
                    + "without a '0x' prefix, of at most 16 hex digits (the "
                    + "leading 16 of a digest are a fine seed, written down "
                    + "as such)",
                repair: seedRepair(program: program))
        }
        return value
    }

    static func seedRepair(program: String) -> String {
        "\(program) experiment evaluate <name> --sample-per-condition <n> "
            + "--sample-seed 0x5eed0a5e5eed0a5e  (any 64-bit value; record it "
            + "in the preregistration — the same seed always draws the same "
            + "records)"
    }

    /// Validate the flag PAIR before anything is read or written.
    ///
    /// `nil` for the full-corpus case (neither flag given), a validated
    /// request when both are, and a malformed-invocation refusal when exactly
    /// one is. Called at the CLI edge AND at the top of the task, so no
    /// caller can reach the draw with half a request.
    public static func resolveRequest(
        samplePerCondition: String?, sampleSeed: String?, program: String
    ) throws -> Request? {
        let sizeText = samplePerCondition?.trimmingCharacters(in: .whitespaces)
        let seedText = sampleSeed?.trimmingCharacters(in: .whitespaces)
        let hasSize = !(sizeText ?? "").isEmpty
        let hasSeed = !(seedText ?? "").isEmpty
        if !hasSize, !hasSeed { return nil }
        if hasSize, !hasSeed {
            throw ExperimentError.malformed(
                "--sample-per-condition \(sizeText ?? "") was given without "
                    + "--sample-seed: a subsample nobody can redraw is not "
                    + "evidence, so the draw refuses rather than choosing a "
                    + "seed for you",
                repair: seedRepair(program: program))
        }
        if hasSeed, !hasSize {
            throw ExperimentError.malformed(
                "--sample-seed \(seedText ?? "") was given without "
                    + "--sample-per-condition: with no sample size the full "
                    + "corpus is coded, and the seed would be stamped on a "
                    + "coding it did not shape",
                repair: "add --sample-per-condition <n> to draw a subsample, "
                    + "or drop --sample-seed and run \(program) experiment "
                    + "evaluate <name> to code the full corpus")
        }
        guard let count = Int(sizeText ?? ""), count >= 1 else {
            throw ExperimentError.malformed(
                "--sample-per-condition must be a whole number of records "
                    + "of at least 1, not '\(sizeText ?? "")' — a subsample "
                    + "of zero records is a design nobody can report",
                repair: "\(program) experiment evaluate <name> "
                    + "--sample-per-condition 2400 --sample-seed "
                    + "<seed>, or drop both flags to code the full corpus")
        }
        return Request(
            samplePerCondition: count,
            seed: try parseSeed(seedText ?? "", program: program))
    }

    // MARK: - The draw

    /// The per-stratum stream seed: the first 8 bytes, big-endian, of
    /// SHA-256(seed as 8-byte big-endian ‖ each length-prefixed UTF-8 part).
    ///
    /// Length-prefixing is not decoration: without it `("ab", "c")` and
    /// `("a", "bc")` would hash identically, so two different cells of one
    /// condition could share a draw — the same reason
    /// `TokenBankDownsampler.corpusHash(texts:)` length-prefixes.
    public static func streamSeed(_ seed: UInt64, _ parts: String...) -> UInt64 {
        streamSeed(seed, parts: parts)
    }

    public static func streamSeed(_ seed: UInt64, parts: [String]) -> UInt64 {
        var hasher = SHA256()
        hasher.update(data: bigEndianBytes(seed))
        for part in parts {
            let encoded = Data(part.utf8)
            hasher.update(data: bigEndianBytes(UInt64(encoded.count)))
            hasher.update(data: encoded)
        }
        return hasher.finalize().prefix(8).reduce(UInt64(0)) {
            ($0 << 8) | UInt64($1)
        }
    }

    private static func bigEndianBytes(_ value: UInt64) -> Data {
        var big = value.bigEndian
        return withUnsafeBytes(of: &big) { Data($0) }
    }

    /// A full Fisher–Yates permutation of `0..<count` over SplitMix64 — the
    /// same loop as `TokenBankDownsampler.selectedIndices`, without the final
    /// sort, because here the ORDER is the answer (which promptID receives
    /// the next remainder record).
    public static func seededOrder(count: Int, seed: UInt64) -> [Int] {
        guard count > 0 else { return [] }
        var rng = SplitMix64(seed: seed)
        var pool = Array(0..<count)
        for position in 0..<count {
            let remaining = count - position
            let offset = Int(rng.next() % UInt64(remaining))
            pool.swapAt(position, position + offset)
        }
        return pool
    }

    /// How many records each promptID contributes: `floor(total / P)` each,
    /// then the remainder handed out one at a time in seeded promptID order,
    /// skipping any promptID already at its population and passing its quota
    /// on.
    ///
    /// The skip is what makes "exactly `total` records, always" an invariant
    /// on a ragged run (a partial or resumed source run need not be
    /// rectangular). It terminates because `select` has already refused when
    /// the condition's total population is below `total`, so headroom always
    /// exists until the last record is placed.
    static func allotments(
        populations: [Int], total: Int, seed: UInt64
    ) -> [Int] {
        let count = populations.count
        guard count > 0, total > 0 else { return Array(repeating: 0, count: count) }
        var allot = populations.map { Swift.min(total / count, $0) }
        let order = seededOrder(count: count, seed: seed)
        var remaining = total - allot.reduce(0, +)
        while remaining > 0 {
            var placed = 0
            for index in order where remaining > 0 {
                if allot[index] < populations[index] {
                    allot[index] += 1
                    remaining -= 1
                    placed += 1
                }
            }
            if placed == 0 { break }  // guarded by the population refusal
        }
        return allot
    }

    /// One record's stratum coordinates, so the draw does not depend on the
    /// concrete record type. `ExperimentTasks.EvaluationGeneration` maps onto
    /// it directly.
    public struct Coordinate: Sendable, Equatable {
        public let condition: String
        public let promptID: String
        public let sampleIndex: UInt64

        public init(condition: String, promptID: String, sampleIndex: UInt64) {
            self.condition = condition
            self.promptID = promptID
            self.sampleIndex = sampleIndex
        }
    }

    /// Codeable records per condition — what an over-ask is measured against.
    public static func population(
        of coordinates: [Coordinate]
    ) -> [String: Int] {
        var counts: [String: Int] = [:]
        for coordinate in coordinates {
            counts[coordinate.condition, default: 0] += 1
        }
        return counts
    }

    /// The source-run positions to KEEP, sorted ascending.
    ///
    /// Refuses — writing nothing, because this runs before the evaluate run
    /// directory is minted — when any condition holds fewer than
    /// `request.samplePerCondition` codeable records.
    public static func selectedPositions(
        _ coordinates: [Coordinate], request: Request, program: String
    ) throws -> [Int] {
        let populations = population(of: coordinates)
        let short = populations
            .filter { $0.value < request.samplePerCondition }
            .sorted {
                $0.value == $1.value
                    ? utf8Precedes($0.key, $1.key) : $0.value < $1.value
            }
        if let (tightest, available) = short.first {
            let named = short.map { "'\($0.key)' has \($0.value)" }
                .joined(separator: ", ")
            throw ExperimentError.malformed(
                "--sample-per-condition \(request.samplePerCondition) exceeds "
                    + "what the source run holds: \(named) codeable "
                    + "record(s). A subsample cannot be larger than the "
                    + "stratum it is drawn from, and clamping it would code a "
                    + "smaller design than the one that was preregistered "
                    + "while every stamp still said "
                    + "\(request.samplePerCondition)",
                repair: "re-run with --sample-per-condition \(available) or "
                    + "less (condition '\(tightest)' is the binding stratum), "
                    + "or drop both sample flags and run \(program) "
                    + "experiment evaluate <name> to code all "
                    + "\(coordinates.count) record(s)")
        }

        var keep: Set<Int> = []
        for condition in populations.keys.sorted(by: utf8Precedes) {
            var cells: [String: [Int]] = [:]
            for (position, coordinate) in coordinates.enumerated()
            where coordinate.condition == condition {
                cells[coordinate.promptID, default: []].append(position)
            }
            let prompts = cells.keys.sorted(by: utf8Precedes)
            for prompt in prompts {
                cells[prompt]?.sort {
                    coordinates[$0].sampleIndex == coordinates[$1].sampleIndex
                        ? $0 < $1
                        : coordinates[$0].sampleIndex
                            < coordinates[$1].sampleIndex
                }
            }
            let allot = allotments(
                populations: prompts.map { cells[$0]?.count ?? 0 },
                total: request.samplePerCondition,
                seed: streamSeed(request.seed, condition))
            for (index, prompt) in prompts.enumerated() {
                let positions = cells[prompt] ?? []
                let chosen = TokenBankDownsampler.selectedIndices(
                    count: positions.count, cap: allot[index],
                    seed: streamSeed(request.seed, condition, prompt))
                for offset in chosen { keep.insert(positions[offset]) }
            }
        }
        return keep.sorted()
    }

    /// The stamp for a completed draw.
    public static func stamp(
        _ request: Request, sampled: Int, source: Int
    ) -> Stamp {
        Stamp(
            rule: rule, samplePerCondition: request.samplePerCondition,
            sampleSeed: request.seedText, sampledRecords: sampled,
            sourceRecords: source)
    }

    /// The human count every line says: `"7200 record(s)"` for a full corpus,
    /// `"2400 of 7200 record(s) (seeded subsample)"` for a sampled one.
    ///
    /// One function so no line can drift into implying a full-corpus coding —
    /// the whole point of the loud stamping is that a reader CANNOT mistake
    /// the two, and a report is read line by line, not block by block.
    public static func codedPhrase(_ stamp: Stamp?, total: Int) -> String {
        guard let stamp else { return "\(total) record(s)" }
        return "\(stamp.sampledRecords) of \(stamp.sourceRecords) record(s) "
            + "(seeded subsample)"
    }

    /// The sample flags on a pairedJudge evaluate.
    ///
    /// Scoped deliberately (2026-08-29): the coding instrument's unit of
    /// analysis is a RECORD, which is what a per-condition stratified draw is
    /// defined over. The paired judge's unit is a PAIR — a variant record
    /// joined to the baseline record of its (promptID, sampleIndex) cell —
    /// where "baseline" is not a sampled condition at all but the other half
    /// of every comparison, so "n records per condition" does not name a set
    /// of pairs. Silently ignoring the flags would be the trap the `--shard`
    /// refusal exists to prevent: a correct-looking command line the CLI
    /// half-executes.
    public static func pairedRefusal(program: String) -> ExperimentError {
        ExperimentError.malformed(
            "--sample-per-condition/--sample-seed apply to the per-response "
                + "coding instrument only: this study's pinned rubric is a "
                + "paired comparison, whose unit is a (baseline, variant) "
                + "PAIR rather than a record, so a per-condition record count "
                + "does not name a set of pairs to judge",
            repair: "drop both sample flags and run \(program) experiment "
                + "evaluate <name> to judge every pair, or pin a "
                + "perResponseCoding rubric if per-record coding is the design")
    }

    /// Byte-order comparison over UTF-8, the cross-engine sort. Swift's `<`
    /// on `String` normalizes; Python's `sorted` does not, and the stratum
    /// order decides which records the draw picks.
    static func utf8Precedes(_ lhs: String, _ rhs: String) -> Bool {
        Array(lhs.utf8).lexicographicallyPrecedes(Array(rhs.utf8))
    }
}
